api/tests/functional-tests/backend/core/test_llm_clients.py

```diff
@@ -0,0 +1,302 @@
+import datetime
+from unittest.mock import MagicMock
+
+# import openai
+import pytest
+
+# from mistralai.exceptions import MistralException
+from mistralai.models.chat_completion import (
+    ChatCompletionResponse,
+    ChatCompletionResponseChoice,
+    ChatMessage,
+    FinishReason,
+)
+from mistralai.models.common import UsageInfo
+from openai.types.chat import ChatCompletionMessage
+from openai.types.chat.chat_completion import ChatCompletion, Choice
+
+from valor_api.backend.core.llm_clients import (
+    LLMClient,
+    WrappedMistralAIClient,
+    WrappedOpenAIClient,
+)
+
+
+def test_LLMClient(monkeypatch):
+    """Check that this parent class mostly throws NotImplementedErrors, since its methods are intended to be overridden by its children."""
+
+    def _return_valid_answer_relevance_response(*args, **kwargs):
+
+        if "generate a list of statements" in args[1][1]["content"]:
+            return """```json
+{
+    "statements": [
+        "statement 1",
+        "statement 2",
+        "statement 3",
+        "statement 4"
+    ]
+}```"""
+        elif (
+            "determine whether each statement is relevant to address the input"
+            in args[1][1]["content"]
+        ):
+            return """```json
+{
+    "verdicts": [
+        {
+            "verdict": "no",
+            "reason": "The statement has nothing to do with the query."
+        },
+        {
+            "verdict": "yes"
+        },
+        {
+            "verdict": "idk"
+        },
+        {
+            "verdict": "yes"
+        }
+    ]
+}```"""
+        else:
+            raise ValueError
+
+    def _return_valid_coherence_response(*args, **kwargs):
+        return "5"
+
+    def _return_invalid_response(*args, **kwargs):
+        return "some bad response"
+
+    client = LLMClient(api_key=None, model_name="model_name")
+
+    fake_message = [{"key": "value"}]
+
+    with pytest.raises(NotImplementedError):
+        client.connect()
+
+    fake_message = client.process_messages(fake_message)
+
+    with pytest.raises(NotImplementedError):
+        client(fake_message)
+
+    # patch __call__ with a valid response
+    monkeypatch.setattr(
+        "valor_api.backend.core.llm_clients.LLMClient.__call__",
+        _return_valid_answer_relevance_response,
+    )
+
+    assert 0.5 == client.answer_relevance("some query", "some answer")
+
+    # patch __call__ with an invalid response
+    monkeypatch.setattr(
+        "valor_api.backend.core.llm_clients.LLMClient.__call__",
+        _return_invalid_response,
```

Comment:
Add another test for if _generate_statements() works as desired but _generate_anwswer_relevance_verdicts() gets a bad llm call